docs: Refactoring the Documentation

taeuscherpferd opened the pull request:
This is the first pull request moving towards a full refactor of the documentation for React-Three-XR. This PR includes the following changes

- Implemented a script to allow the documentation to be auto generated from the JSDoc in the code `npm run generate_docs`
- Migrated all of the existing documentation to JSDoc comments on the components and hooks
- Fixed spelling and grammar errors as I came across them
- Added a couple of better descriptions as I came across them

In the future I will be going through and updating each component individually with helpful descriptions, code examples, and links to tutorials that showcase the components use.

I understand that this is a large change and am happy to keep working with it until we get it right. I just wanted to get the pull request up so that if I'm headed down the wrong path, we can catch it, and course correct. 

Worth noting as well that I have not yet been able to preview how this markdown will look rendered on the docs page. May need to ask Abernier how to do that.

bbohlender commented:
Awesome. Looks great. Yeah, a preview would be important.

Maybe @abernier would also like to give his general take on the approach :)

Another question: What are you using to re-order the dependencies?

abernier commented:
iiuc, markdown files are auto-generated

then, should probably better not to commit them, but generate them by launching the script: in CI or pre-commit hook or whatever...

wdyt?

taeuscherpferd commented:
@abernier Yes you are correct, there is now a script to auto generate the markdown based off of the JSDoc comments in the code. 

Totally agree about generating them in the pipeline or in a hook, but am still not super familiar with the build process for the project, and wasn't sure how exactly to go about it. Would you prefer in the CI or in a pre-commit hook? I think CI makes sense, but am happy to do whichever you think is best  👍

The main issue that we're running into though, is that we would like to preview what these changes to the markdown files will end up looking like before pushing to prod. Do you have a staging or dev server/area for the documentation setup that we could place the markdown files in to see how they turn out?

Thanks for your help!

taeuscherpferd commented:
@bbohlender I'm using VSCode's built in `organizeImports` functionality on save. It's nice because it ensures that you never leave an unused import in a file, but can also be a pain the first time in a file because it shuffles everything around. This is the option in VSCode's settings.json

```json
  "editor.codeActionsOnSave": {
    "source.organizeImports": "explicit"
  },
  ```
  
  I can unshuffle them if you prefer. My only thoughts are it's nice to not have to worry about unused imports, and the order of dependency imports doesn't affect how the code functions anyway

taeuscherpferd commented:
@bbohlender I'd be happy to look into that. :) Just to be clear, you want that as a separate pull request? In the meantime what would you like done with this one?

bbohlender commented:
@taeuscherpferd yes, a seperate PR would be nice for that since I assume it would be a lot of changes and we can directly accept them without reviewing every single file since its just a eslint fix :)

regarding next steps, a CICD integration would be nice to generate the .md files when building the docs and adding them to the .gitignore. And also previewing would be necassary to confirm this works. I was following the instructions for the docs to preview them locally and it worked, even though not really fast with docker and everything

Once those things are done, let's get this merged fast and do this:
> In the future I will be going through and updating each component individually with helpful descriptions, code examples, and links to tutorials that showcase the components use.

in a seperate PR, wdyt? :)

bbohlender commented:
For me, previewing it locally would be enough, but If you prefer we can set it up as a CICD for e.g. a "docs" branch :)

Maybe I can give you permission to directly push to the docs branch, so you can deploy a preview yourself and get feedback from others in a fast cycle, wdyt?

I think there are many people interested in giving feedback

taeuscherpferd commented:
Just an update here. 

I got everything running locally, found out that the docs use mdx, fixed a bunch of errors, and.......... it's super ugly 😂😭

I'm reworking the mdx to see what I can do to make it look a little better, should be done with that by tomorrow 👍

Not sure where to host a public preview, maybe I can just throw it up on a bucket or something and let anyone who wants to chime in have a look at it. If the docs branch is already setup for that, then sounds good to me. Honestly a CICD would be nice, but I don't think it'd be worth the hassle at the moment 👍

bbohlender commented:
awesome, let's just merge and test in prod? :D we can always revert or do you think its to big of a change and people would be upset? :)

taeuscherpferd commented:
Haha well you know what they say, real men test in Prod ;) XD Not a bad idea, give me another day to pretty it up a bit, and then I feel pretty confident that we could send it through. 👍 I'll ping you when it's ready to go

bbohlender commented:
awesome, looking forward to it! Great work!

taeuscherpferd commented:
@bbohlender Alright! This should be good to go whenever you want to give it another review/push it through 👍

bbohlender reviewed (commented):
sorrry forgot to submit the review

taeuscherpferd commented on packages/react/xr/src/hooks.ts:
```diff
@@ -1,12 +1,15 @@
-
-export function useHover(ref: RefObject<Object3D | null>): boolean
-
-export function useHover(ref: RefObject<Object3D | null>, onChange: (hover: boolean, event: PointerEvent) => void): void
```
It's definitely not my intentention to break anything. The parser has a hard time dealing with code that is overloaded like this. In other cases I've worked around it but looking at the function as it is, with onChange being optional anyway, I did not see the overloads actually adding anything to the function. I'm still relatively new to package development though, so I can definitely see how I might be wrong here, I will restore these 👍

taeuscherpferd replied:
As an extra precaution, I've also done another pass and removed all non comment changes in all of the files except for adding a shared interface to the render/show guards as it helped with the parser, and I figured that those functions are more or less 'paired' and should share props anyway. If you'd like though, I can revert that as well. 👍

taeuscherpferd commented on packages/react/xr/package.json:
```diff
@@ -32,7 +32,10 @@
   "devDependencies": {
-    "@react-three/fiber": "rc"
+    "@react-three/fiber": "rc",
+    "typedoc": "^0.28.3",
```
My only reasoning is that we were only doing the docs for the XR package at the moment. I didn't see it being a very useful package in the majority of the projects in the repo (i.e. I don't think that we'll be generating documentation fo the examples), so I figured I would keep it local to the packages that I thought would use it. If you feel that it's better suited top level though, I can move it 👍

The change in the handle package.json slipped past me, and I've removed it.

bbohlender replied:
at some point we should also document stuff in react-three/handle, therefore it might be a good idea, but not a blocker for this PR

bbohlender commented:
@taeuscherpferd let me know when you adressed everything (maybe you already did) and I'll merge :)

taeuscherpferd commented:
@bbohlender I have addressed all of your feedback, thanks again for doing such a good review. We could merge it through, but now I'm getting cold feet after the mention of a breaking change. 😅 I want to make sure that these are changes that you want, so I am considering trying to put some sort of preview together again. However, if you do want to push it through to prod, have a look there and then decide if you want to keep it, or roll it back for more changes, that works too. Just want to make sure you're getting quality work. 👍

bbohlender commented:
@taeuscherpferd seems like the API folder is not picked up in the deployed version

taeuscherpferd commented:
@bbohlender You're correct :( Hmm something must have gone wrong with the build step....

bbohlender commented:
@taeuscherpferd why do you assume that? Couldn't it just be that the docs doesnt support this deep nesting?

taeuscherpferd commented:
@bbohlender I actually didn't do the nesting structure due to previous concerns. The API folder should actually sit top level with `Getting Started` and `Tutorials`. My only other theory is that because the API folder is in the .gitignore, the docs step might have skipped it

bbohlender commented:
I think you added the generate api docs step to the examples workflow and not to the docs workflow :D

taeuscherpferd commented:
Are they not the same? If I did mess it up, my bad 😅I'm not much of a devops guy, but it looked to me like the examples were being built as part of the docs step to publish. Is the correct place just in the `packages.yml`?

taeuscherpferd commented:
Ah I think I see what it might be. There doesn't look to be a `pnpm install` step. I wonder if that's causing the API build script not to complete (weird that the examples build without installing dependencies though)

taeuscherpferd commented:
Ah I think I see it now. I thought that the docs step was just prepping the docs project to receive the files from the publish step, but it's actually pointing directly at the repo itself to pull in the docs. This might be tricky to work around as the API folder doesn't actually get committed to source.

taeuscherpferd commented:
![image](https://github.com/user-attachments/assets/b1a4fddd-1f8a-4e4a-b209-24ea6dc0a523)

@abernier Looking at the docs build script it looks like it is setup to pull the docs from the github repo directly. If that's the case then it seems like our generation script would have to commit the generated files to source in order for them to be picked up by the docs server. Is there a way to maybe inject an artifact into the build process to add another folder? I'm thinking probably not, as it would make sense for the docs build to point at the repo directly. 

@bbohlender Would you like me to create another pull request to add the API folder to source, or just roll back for now?

bbohlender commented:
@abernier Could you add an option to disable checking out the repo, which would allow us to checkout the repo ourselves and run the generate scripts before the "build docs" workflow is run?

abernier commented:
i'd prefer not to, bit you can execute the `docker run` command manually if you want, this is basically the only purpose of the shared workflow
